### PR TITLE
add implementation to save files

### LINKED ISSUE #8: Offer a dialog to save files as well
While maybe not strictly a "picker" anymore, it would be nice to also have a dialog for saving files.

### PR DESCRIPTION
**Pull request by ghost**
Addresses Issue https://github.com/Wavesonics/compose-multiplatform-file-picker/issues/8

Milestones:
- [x] Android
- [x] ~~Web~~ The Web implementation was removed during a refactor due to not being able to implement it.
- [x] JVM
- [x] iOS
- [ ] macOS
    - [x] v1
    - [ ] Add support for more customization e.g. `prompt`, `message`, `nameFieldLabel` 
    - [ ] Testing

### CODE REVIEW

**Comment by santiwanti** [comment]
@shalva97 @Wavesonics I got save working and tested on all platforms except macOS. On macOS it is implemented, but I haven't been able to test it for some reason. Running the example compiles it successfully, but it doesn't launch. Could it be because I'm running on Arm and not x64?

I will keep trying to get macOS working and tested. In the meantime any info on whether this feature is wanted in the library or if it should be kept in a fork or any feedback on the implementation details would be greatly appreciated.

Cheers

**Comment by akriese** [comment]
Hi @santiwanti, I am also interested in the saveFilePicker you are implementing. I am wondering if the API youre implementing currently fits the other functions the package offers. Specifically, I would say, the picker should not handle the saving of the contents itself. It should instead take an onFilePicked-function as input which would be provided by the user where the user decides how to work with the selected file.
I'd like to contribute to this to push this PR through, so please tell me, if I can help you in any way :)

**Comment by santiwanti** [comment]
@akriese that's a good point. I'm refactoring the code now so that the callback is the same as for the other functions and also refactoring the name from `SaveFilePicker` to `NewFilePicker` although I don't love this name because the user could select an existing file.

As for contributing I will implement these changes today and after that the only things left will be to test it on MacOS which I haven't been able to do and getting feedback from maintainers to see if anything should be changed. I will ping you once it's ready to test on MacOS.

**Comment by akriese** [comment]
A colleague of mine uses macOS (x86, I guess). I could let him test your code, if you give us instructions on how to test it. Can one simply clone the repo and run the example apps?

**Comment by akriese** [comment]
I personally think `SaveFilePicker` is a good name, as that is usually what you're doing with the picker. Also, it is consistent with the `TinyFileDialogs.tinyfd_saveFileDialog` you're using, so I dont see a reason to rename it to `NewFilePicker` :)

With other filepickers it is similar, that the programmer (user of the package) has to handle file overwrites. At least thats how you have to do it with Swing's `JFileChooser().showSaveDialog()` too.

**Comment by santiwanti** [comment]
I have refactored the code so that the user can choose to do once the new file has been created.

The Web code had to be removed and now throws a NotImplemented Exception because there is no way to implement it currently.

Besides this everything else is the same.

The UX that is a bit strange is the iOS UX because the way it is implemented is that a temporary file is created and then the user can decide where to move the file. That means that the file picker action the user sees is "Move" which is a bit deceptive.

Lastly @akriese the Mac code has still not been tested because I can't run it for some reason. To test the app you have to go through the instructions laid out here: https://www.jetbrains.com/help/kotlin-multiplatform-dev/compose-multiplatform-getting-started.html and after that it's as easy as running the code.

**Comment by akriese** [comment]
Hey, sorry for letting you wait. I just tried the jvm example. The example is missing the `showSaveFile = false` in the callback body but otherwise it seems to work fine.
I'll ask my friend now to try it on Mac and will report back hen we have a result.

**Comment by akriese** [comment]
So, on his mac the jvm version works, but sometimes (we cant really tell when), picking an existing file will cut off the file extension of that file.
The macOS version does not run at all, as it was the case for you. It compiles and when running it, it says the execution is finished without a window popping up. He is using an ARM mac too though, so we cant exclude the possibility of that being a problem...

**Comment by shalva97** [comment]
thanks for all the work. I think it will be okay to keep both file picker dialogs and save file dialogs in this library. 

First I'd like to merge https://github.com/Wavesonics/compose-multiplatform-file-picker/pull/89, it will cause, but should be easier after that.

Regarding MacOS, it needs one more compilation target macosArm64 to be added. Anyways, I have MacOS with intel so I can test it on macosX64.

**Comment by santiwanti** [comment]
@shalva97 I didn't want to add the macosArm64 target given that it wasn't already in the project. If you think adding the target is a good idea I would be willing to add it in another PR and test/fix all the logic. In the meantime if you can test the macos saving code that would be great. I can fix the bugs that come up and we can find a way to collaborate more synchronously if it's not working. 
I will keep an eye out on https://github.com/Wavesonics/compose-multiplatform-file-picker/pull/89 and update this PR once that PR is merged.

**Comment by shalva97** [comment]
Hi. Do you want to be a maintainer? I will ask @Wavesonics, he usually responds on Kotlin's Slack

Personally I dont use this library, there was just my pet project, which used it but I have abandoned that.

**Comment by akriese** [comment]
Hey @santiwanti, I just noticed that #89 was merged. Could you fix conflicts so @shalva97 can merge this at some point? :)

**Comment by santiwanti** [comment]
Sorry I was afk for a few days. 

@shalva97 I would be down to being a maintainer, but lets talk about it more and with @Wavesonics to make sure we are on the same page and don't step on eachothers toes. You can message me on Slack @ swanti.

@akriese I will be back in full force tomorrow and will fix conflicts and push the changes.

**Comment by santiwanti** [comment]
@shalva97 @Wavesonics this still has to be tested on an Intel Mac. I haven't been able to test it myself

**Comment by santiwanti** [comment]
I will be working on https://github.com/Wavesonics/compose-multiplatform-file-picker/issues/113 so that I can test the MacOS changes. If someone can test the save file picker on an intel mac before that and confirm that it works I will merge this PR. If not, I will merge it after https://github.com/Wavesonics/compose-multiplatform-file-picker/issues/113 is merged and I can test the save file picker on mac myself

File: mpfilepicker/src/macosX64Main/kotlin/com/darkrockstudios/libraries/mpfilepicker/FilePicker.macos.kt
Comment (shalva97) [inline mpfilepicker/src/macosX64Main/kotlin/com/darkrockstudios/libraries/mpfilepicker/FilePicker.macos.kt:141]: Message is the title of the dialog. Don't set fileExtension to it.

```suggestion
				if (fileExtension != null) message = title
```

File: mpfilepicker/src/macosX64Main/kotlin/com/darkrockstudios/libraries/mpfilepicker/FilePicker.macos.kt
Comment (shalva97) [inline mpfilepicker/src/macosX64Main/kotlin/com/darkrockstudios/libraries/mpfilepicker/FilePicker.macos.kt:134]: This comment does not really add any information. I think it should be removed
```suggestion
```

File: examples/macosX64/src/macosX64Main/kotlin/main.kt
Comment (shalva97) [inline examples/macosX64/src/macosX64Main/kotlin/main.kt:124]: This does not work for me. `NSFileHandle.fileHandleForWritingAtPath(filePath)` returns null.

Comment (shalva97) [reply examples/macosX64/src/macosX64Main/kotlin/main.kt:124]: To keep examples simpler, try https://github.com/Kotlin/kotlinx-io, it has common code, which can be used to save file on any platform. But anyway `NSFileHandle` is also okay.

Something like:
```kotlin
import io.ktor.utils.io.core.*
import kotlinx.io.buffered
import kotlinx.io.files.Path
import kotlinx.io.files.SystemFileSystem
import kotlinx.io.write
import kotlin.test.Test

class Learning {
    @OptIn(ExperimentalStdlibApi::class)
    @Test
    fun saveStuffToFile() {
        val file = Path("./test")

        SystemFileSystem.sink(file).buffered().use {
            it.write("1234".toByteArray())
        }
    }
}
```

File: examples/jvm/src/main/kotlin/Main.kt
Comment (shalva97) [inline examples/jvm/src/main/kotlin/Main.kt:92]:
```suggestion
		} ?: false
		showSaveFile = false
```

Comment (santiwanti) [reply examples/jvm/src/main/kotlin/Main.kt:92]: I can't open any dialog more than once. that is actually something I wanted to take a look at to see if it's our code or if it's the example compose code that does something weird.

**Comment by zhelenskiy** [comment]
I'm interested in the feature, and I have x64 win 11, Linux (Kubuntu) and Mac, and ARM Mac. If you need any of the platforms, I can check it.

File: examples/macosX64/src/macosX64Main/kotlin/main.kt
Comment (santiwanti) [reply examples/macosX64/src/macosX64Main/kotlin/main.kt:124]: I have updated the example, but again I can't test it. Feel free to test it and merge it. If the path is returned "our job is done", although it would be great if the example worked properly.

File: examples/jvm/src/main/kotlin/Main.kt
Comment (santiwanti) [reply examples/jvm/src/main/kotlin/Main.kt:92]: I have fixed this. You should be able to open the dialog more than once now

**Comment by shalva97** [comment]
can confirm, creating and writing some nice text in a file did work. 
Screenshot 2024-02-23 at 23 23 41

File: examples/macosX64/src/macosX64Main/kotlin/main.kt
Comment (shalva97) [reply examples/macosX64/src/macosX64Main/kotlin/main.kt]: just tested and it works 🚀

File: examples/jvm/src/main/kotlin/Main.kt
Comment (shalva97) [reply examples/jvm/src/main/kotlin/Main.kt:92]: nice

**Review by shalva97 — APPROVED** [review]

**Comment by zhelenskiy** [comment]
Are there any updates?

**Comment by zhelenskiy** [comment]
@santiwanti @Wavesonics are there any updates? Can I somehow help you? Is there any ETA?

**Pull request closed without merging**
